zerver/lib/actions.py

```diff
@@ -3050,9 +3050,11 @@ def do_change_plan_type(user: UserProfile, plan_type: int) -> None:
 
     if plan_type == Realm.PREMIUM:
         realm.max_invites = Realm.MAX_INVITES_PREMIUM
+        realm.message_visibility_limit = None
     elif plan_type == Realm.LIMITED:
         realm.max_invites = settings.INVITES_DEFAULT_REALM_DAILY_MAX
-    realm.save(update_fields=['_max_invites'])
+        realm.message_visibility_limit = 10000
+    realm.save(update_fields=['_max_invites', 'message_visibility_limit'])
```

Comment:
This logic is wrong; we should be setting `realm.message_visibility_limit` to `None` whenever `plan_type` is changed away from `Realm.LIMITED`, regardless of what the other plan type is.  Otherwise when we mark an open source org as PREMIUM_FREE, the wrong thing will happen.

Reply:
A possible structure for this function might be
* A bunch of if statements on the `old_value`, to "unset" values
* A bunch of if statements on `plan_type`, to set values

That way we won't have to worry about the `n^2` different possibilities.